src/pam/main.cc

```diff
@@ -0,0 +1,394 @@
+  // zombie process)
+  optional_task<int> child_task(packaged_task<int()>([&] {
+    int status;
+    wait(&status);
+    {
+      unique_lock<mutex> lk(m);
+      confirmation_type = Type::Howdy;
+    }
+    cv.notify_all();
+    return status;
+  }));
+  child_task.activate();
+
+  // This task waits for the password input (if the workaround wants it)
+  optional_task<tuple<int, char *>> pass_task(
+      packaged_task<tuple<int, char *>()>([&] {
+        char *auth_tok_ptr = nullptr;
+        int pam_res = pam_get_authtok(pamh, PAM_AUTHTOK,
+                                      (const char **)&auth_tok_ptr, nullptr);
+        {
+          unique_lock<mutex> lk(m);
+          confirmation_type = Type::Pam;
+        }
+        cv.notify_all();
+        return tuple<int, char *>(pam_res, auth_tok_ptr);
+      }));
+
+  if (auth_tok) {
+    pass_task.activate();
+  }
+
+  // Wait for the end either of the child or the password input
+  {
+    unique_lock<mutex> lk(m);
+    cv.wait(lk);
+    final_type = confirmation_type;
+  }
+
+  if (final_type == Type::Howdy) {
+    // We need to be sure that we're not going to block forever if the
+    // child has a problem
+    if (child_task.wait(3s) == future_status::timeout) {
+      kill(child_pid, SIGTERM);
+    }
+    child_task.stop(false);
+
+    // If the workaround is native
+    if (auth_tok) {
+      // We cancel the thread using pthread, pam_get_authtok seems to be a
+      // cancellation point
+      if (pass_task.is_active()) {
+        pass_task.stop(true);
+      }
+    }
+    int howdy_status = child_task.get();
+
+    // If exited successfully
+    if (howdy_status == 0) {
+      if (!reader.GetBoolean("core", "no_confirmation", true)) {
+        // Construct confirmation text from i18n string
+        string confirm_text = dgettext("pam", "Identified face as {}");
+        string identify_msg(
+            confirm_text.replace(confirm_text.find("{}"), 2, string(user_ptr)));
+        // Send confirmation message to user
+        conv_function(PAM_TEXT_INFO, identify_msg.c_str());
+      }
+      syslog(LOG_INFO, "Login approved");
+      return PAM_SUCCESS;
+    } else {
+      return on_howdy_auth(howdy_status, conv_function);
+    }
+  }
+
+  // The branch with Howdy confirmation type returns early, so we don't need an
+  // else statement
+
+  // Again, we need to be sure that we're not going to block forever if the
+  // child has a problem
+  if (child_task.wait(1.5s) == future_status::timeout) {
+    kill(child_pid, SIGTERM);
+  }
+  child_task.stop(false);
+
+  // We just wait for the thread to stop since it's this one which sent us the
+  // confirmation type
+  if (workaround == Workaround::Input && auth_tok) {
+    pass_task.stop(false);
+  }
+
+  char *token = nullptr;
+  tie(pam_res, token) = pass_task.get();
+
+  if (pam_res != PAM_SUCCESS)
+    return pam_res;
+
+  int howdy_status = child_task.get();
```

Comment:
What will this value be when the child task is killed due to timeout? Seems like undefined behaviour.

Reply:
IIRC, it will be the signal which terminated it, something like SIGKILL, so definitely not a success.